packages/lib/src/actionMiddlewares/draftMiddleware.ts

```diff
@@ -0,0 +1,102 @@
+import { getParentToChildPath, resolvePath } from "../parent"
+import { applyPatches, onPatches, OnPatchesDisposer, patchRecorder } from "../patch"
+import { isRootStore, registerRootStore, unregisterRootStore } from "../rootStore"
+import { clone } from "../snapshot"
+import { assertTweakedObject } from "../tweaker/core"
+
+/**
+ * Callback function for `DraftManager.withDraft`.
+ */
+export type WithDraftCallback<T extends object> = (draftNode: T) => boolean | undefined | void
+
+/**
+ * Manager class returned by `draftMiddleware` that allows you to make changes to a draft copy of
+ * the original subtree and apply them to the original subtree or reject them.
+ */
+export class DraftManager {
+  /**
+   * The draft copy of the original subtree.
+   */
+  private readonly subtreeRootDraft: object
+
+  /**
+   * The internal `onPatches` disposer.
+   */
+  private readonly disposer: OnPatchesDisposer
+
+  /**
+   * Creates an instance of `DraftManager`.
+   * Do not use directly, use `draftMiddleware` instead.
+   *
+   * @param subtreeRoot Subtree root target object.
+   */
+  public constructor(private readonly subtreeRoot: object) {
+    assertTweakedObject(subtreeRoot, "subtreeRoot")
+    this.subtreeRootDraft = clone(subtreeRoot)
+    if (isRootStore(subtreeRoot)) {
+      registerRootStore(this.subtreeRootDraft)
+    }
+    this.disposer = onPatches(subtreeRoot, patches => applyPatches(this.subtreeRootDraft, patches))
+  }
+
+  /**
+   * Executes `fn` with a draft of `node`. The changes made to the draft in `fn` can be accepted,
+   * i.e. applied to the original subtree, or rejected.
+   *
+   * @typeparam T Object type.
+   * @param node Object for which to obtain a draft copy.
+   * @param fn Function that is caled with a draft copy of `node`. Any changes made to the draft are
+   * applied to the original subtree when `fn` returns `true`. When `fn` returns `false`,
+   * `undefined` or nothing, the changes made to the draft are rejected.
+   */
+  public withDraft<T extends object>(node: T, fn: WithDraftCallback<T>): void {
+    assertTweakedObject(node, "node")
+
+    const path = getParentToChildPath(this.subtreeRoot, node)
+    if (!path) {
+      return
+    }
+
+    const draftNode = resolvePath<T>(this.subtreeRootDraft, path).value
+    if (!draftNode) {
+      return
+    }
+
+    const recorder = patchRecorder(this.subtreeRootDraft)
+    const accept = fn(draftNode)
+    recorder.dispose()
+
+    if (accept) {
+      applyPatches(
+        this.subtreeRoot,
+        recorder.events.flatMap(event => event.patches)
+      )
+    } else {
+      applyPatches(
+        this.subtreeRootDraft,
+        recorder.events.flatMap(event => event.inversePatches)
+      )
+    }
+  }
+
+  /**
+   * Disposes the draft middleware.
+   */
+  public dispose(): void {
+    this.disposer()
+    if (isRootStore(this.subtreeRootDraft)) {
```

Comment:
remember to dispose of the reaction here if done in a reactive way (as suggested above) while keeping this code too

Reply:
Done, but also see my above comment about the approach I've taken in general.

Reply:
LGTM

Reply:
Done in d21afe4e27488cef866e624c117c67f53db52444. The disposer closure becomes a noop after it has been called once.